src/java/org/jivesoftware/openfire/server/LocalOutgoingServerProxy.java

```diff
@@ -0,0 +1,176 @@
+    private static ExecutorService createPool() {
+        // Create a pool of threads that will process queued session requests.
+        int maxThreads = JiveGlobals.getIntProperty(ConnectionSettings.Server.QUEUE_MAX_THREADS, 20);
+        if (maxThreads < 10) {
+            // Ensure that the max number of threads in the pool is at least 10
+            maxThreads = 10;
+        }
+        ExecutorService pool = Executors.newFixedThreadPool(maxThreads);
+        return pool;
+    }
+    
+    public LocalOutgoingServerProxy(final JID domain) {
+        this.domain = domain;
+        this.session = null;
+        this.packets = null;
+    }
+
+    public LocalOutgoingServerProxy(final String domain) {
+        this.domain = new JID(domain);
+        this.session = null;
+        this.packets = null;
+    }
+
+    public LocalOutgoingServerProxy(final JID domain, ServerSession session) {
+        this.domain = domain;
+        this.session = null;
+        this.packets = null;
+    }
+
+    public LocalOutgoingServerProxy(final String domain, ServerSession session) {
+        this.domain = new JID(domain);
+        this.session = null;
+        this.packets = null;
+    }
+    
+    private static int active = 0;
+    
+    protected synchronized void started() {
+        active = active + 1;
+        log.debug("Now +{} running connection attempts", active);
+    }
+    protected synchronized void finished() {
+        active = active - 1;
+        log.debug("Now -{} running connection attempts", active);
+    }
+
+    /* (non-Javadoc)
+     * @see org.jivesoftware.openfire.ChannelHandler#process(org.xmpp.packet.Packet)
+     */
+    @Override
+    public synchronized void process(final Packet packet) throws UnauthorizedException,
+            PacketException {
+        if (this.session != null) {
+            this.session.process(packet);
+        } else {
+            if (packets == null) {
+                packets = new LinkedBlockingQueue<Packet>();
+                log.info("Queued packet for {}.", domain.toString());
+            }
+            packets.add(packet.createCopy());
+            if (isTrying == false) {
+                final String fromDomain = packet.getFrom().getDomain().toString();
+                final String toDomain = packet.getTo().getDomain().toString();
+                if ((failureTimestamp == -1) || ((System.currentTimeMillis() - failureTimestamp) >= 5000)) {
+                    isTrying = true;
+                    log.debug("Spinning up new session to {}", domain.toString());
+                    pool.execute(new Runnable() {
+                        public void run() {
+                            log.debug("Initiating connection thread for {} -> {} ({})", fromDomain, toDomain, domain.toString());
+                            started();
+                            try {
+                                ServerSession s = LocalOutgoingServerSession.authenticateDomain(fromDomain, toDomain); // Long-running.
```

Comment:
Hmm, Code and comment in the same line is at least controversial in the Java community.